### ISSUE TITLE
Illegal instance declarations

### ISSUE THREAD

**Issue by Fuuzetsu** [pxqr/bittorrent#7]
This is with current HEAD.

```
[ 1 of 35] Compiling Paths_bittorrent ( dist/build/autogen/Paths_bittorrent.hs, dist/build/Paths_bittorrent.o )
[ 2 of 35] Compiling Network.BitTorrent.Internal.Progress ( src/Network/BitTorrent/Internal/Progress.hs, dist/build/Network/BitTorrent/Internal/Progress.o )
Loading package ghc-prim ... linking ... done.
Loading package integer-gmp ... linking ... done.
Loading package base ... linking ... done.
Loading package split-channel-0.2.0.1 ... linking ... done.
Loading package transformers-0.3.0.0 ... linking ... done.
Loading package mtl-2.1.2 ... linking ... done.
Loading package array-0.5.0.0 ... linking ... done.
Loading package deepseq-1.3.0.2 ... linking ... done.
Loading package old-locale-1.0.0.6 ... linking ... done.
Loading package time-1.4.2 ... linking ... done.
Loading package random-1.0.1.1 ... linking ... done.
Loading package MonadRandom-0.1.13 ... linking ... done.
Loading package random-shuffle-0.0.4 ... linking ... done.
Loading package pretty-1.1.1.1 ... linking ... done.
Loading package pretty-class-0.1 ... linking ... done.
Loading package bytestring-0.10.4.0 ... linking ... done.
Loading package mmap-0.5.9 ... linking ... done.
Loading package stm-2.4.3 ... linking ... done.
Loading package async-2.0.1.5 ... linking ... done.
Loading package transformers-base-0.4.2 ... linking ... done.
Loading package monad-control-0.3.3.0 ... linking ... done.
Loading package lifted-base-0.2.3.0 ... linking ... done.
Loading package lifted-async-0.2.0 ... linking ... done.
Loading package containers-0.5.5.1 ... linking ... done.
Loading package text-1.1.1.3 ... linking ... done.
Loading package hashable-1.2.2.0 ... linking ... done.
Loading package scientific-0.3.3.0 ... linking ... done.
Loading package attoparsec-0.12.1.0 ... linking ... done.
Loading package dlist-0.7.1 ... linking ... done.
Loading package syb-0.4.2 ... linking ... done.
Loading package template-haskell ... linking ... done.
Loading package unordered-containers-0.2.5.0 ... linking ... done.
Loading package primitive-0.5.3.0 ... linking ... done.
Loading package vector-0.10.11.0 ... linking ... done.
Loading package aeson-0.7.0.6 ... linking ... done.
Loading package transformers-compat-0.3.3 ... linking ... done.
Loading package contravariant-0.6 ... linking ... done.
Loading package tagged-0.7.2 ... linking ... done.
Loading package distributive-0.4.4 ... linking ... done.
Loading package nats-0.2 ... linking ... done.
Loading package semigroups-0.15.1 ... linking ... done.
Loading package comonad-4.2 ... linking ... done.
Loading package semigroupoids-4.0.2.1 ... linking ... done.
Loading package bifunctors-4.1.1.1 ... linking ... done.
Loading package exceptions-0.6.1 ... linking ... done.
Loading package filepath-1.3.0.2 ... linking ... done.
Loading package prelude-extras-0.4 ... linking ... done.
Loading package profunctors-4.0.4 ... linking ... done.
Loading package free-4.9 ... linking ... done.
Loading package parallel-3.2.0.4 ... linking ... done.
Loading package reflection-1.5.1 ... linking ... done.
Loading package split-0.2.2 ... linking ... done.
Loading package void-0.6.1 ... linking ... done.
Loading package zlib-0.5.4.1 ... linking ... done.
Loading package lens-4.3 ... linking ... done.
Loading package bencoding-0.4.3.0 ... linking ... done.
Loading package data-default-class-0.0.1 ... linking ... done.
Loading package blaze-builder-0.3.3.2 ... linking ... done.
Loading package mmorph-1.0.3 ... linking ... done.
Loading package resourcet-1.1.2.2 ... linking ... done.
Loading package conduit-1.1.7 ... linking ... done.
Loading package unix-2.7.0.1 ... linking ... done.
Loading package directory-1.2.1.0 ... linking ... done.
Loading package parsec-3.1.5 ... linking ... done.
Loading package network-2.5.0.0 ... linking ... done.
Loading package process-1.2.0.0 ... linking ... done.
Loading package streaming-commons-0.1.4.1 ... linking ... done.
Loading package conduit-extra-1.1.2 ... linking ... done.
Loading package fast-logger-2.1.5 ... linking ... done.
Loading package monad-loops-0.4.2.1 ... linking ... done.
Loading package stm-chans-3.0.0.2 ... linking ... done.
Loading package monad-logger-0.3.7 ... linking ... done.
Loading package krpc-0.6.1.0 ... linking ... done.
Loading package appar-0.1.4 ... linking ... done.
Loading package byteorder-1.0.4 ... linking ... done.
Loading package iproute-1.3.0 ... linking ... done.
Loading package bits-extras-0.1.3 ... linking ... done.
Loading package intset-0.1.1.0 ... linking ... done.
Loading package base64-bytestring-1.0.0.1 ... linking ... done.
Loading package case-insensitive-1.2.0.0 ... linking ... done.
Loading package data-default-instances-base-0.0.1 ... linking ... done.
Loading package data-default-instances-containers-0.0.1 ... linking ... done.
Loading package data-default-instances-dlist-0.0.1 ... linking ... done.
Loading package data-default-instances-old-locale-0.0.1 ... linking ... done.
Loading package data-default-0.5.3 ... linking ... done.
Loading package cookie-0.4.1.1 ... linking ... done.
Loading package http-types-0.8.5 ... linking ... done.
Loading package mime-types-0.1.0.4 ... linking ... done.
Loading package cereal-0.4.0.1 ... linking ... done.
Loading package utf8-string-0.3.8 ... linking ... done.
Loading package publicsuffixlist-0.1 ... linking ... done.
Loading package http-client-0.3.6 ... linking ... done.
Loading package byteable-0.1.1 ... linking ... done.
Loading package securemem-0.1.3 ... linking ... done.
Loading package crypto-cipher-types-0.0.9 ... linking ... done.
Loading package cipher-aes-0.2.8 ... linking ... done.
Loading package crypto-random-0.0.7 ... linking ... done.
Loading package cprng-aes-0.5.2 ... linking ... done.
Loading package socks-0.5.4 ... linking ... done.
Loading package asn1-types-0.2.3 ... linking ... done.
Loading package asn1-encoding-0.8.1.3 ... linking ... done.
Loading package cipher-des-0.0.6 ... linking ... done.
Loading package cipher-rc4-0.1.4 ... linking ... done.
Loading package crypto-numbers-0.2.3 ... linking ... done.
Loading package crypto-pubkey-types-0.4.2.2 ... linking ... done.
Loading package cryptohash-0.11.6 ... linking ... done.
Loading package crypto-pubkey-0.2.4 ... linking ... done.
Loading package asn1-parse-0.8.1 ... linking ... done.
Loading package pem-0.2.2 ... linking ... done.
Loading package x509-1.4.11 ... linking ... done.
Loading package x509-store-1.4.4 ... linking ... done.
Loading package x509-validation-1.5.0 ... linking ... done.
Loading package tls-1.2.8 ... linking ... done.
Loading package x509-system-1.4.5 ... linking ... done.
Loading package connection-0.2.3 ... linking ... done.
Loading package http-client-tls-0.2.2 ... linking ... done.
Loading package http-conduit-2.1.4 ... linking ... done.
Loading package entropy-0.3.2 ... linking ... done.
Loading package old-time-1.1.0.2 ... linking ... done.
Loading package convertible-1.1.0.0 ... linking ... done.
Loading package cereal-conduit-0.7.2.2 ... linking ... done.
Loading package cpu-0.1.2 ... linking ... done.
Loading package base32-bytestring-0.2.1.0 ... linking ... done.
Loading package base16-bytestring-0.1.1.6 ... linking ... done.
Loading package SafeSemaphore-0.10.1 ... linking ... done.
Loading package PSQueue-1.1 ... linking ... done.

src/Network/BitTorrent/Internal/Progress.hs:40:1: Warning:
    The import of ‘Data.ByteString.Lazy.Builder.ASCII’ is redundant
      except perhaps to import instances from ‘Data.ByteString.Lazy.Builder.ASCII’
    To import instances alone, use: import Data.ByteString.Lazy.Builder.ASCII()
[ 3 of 35] Compiling Network.BitTorrent.Internal.Cache ( src/Network/BitTorrent/Internal/Cache.hs, dist/build/Network/BitTorrent/Internal/Cache.o )

src/Network/BitTorrent/Internal/Cache.hs:82:19: Warning:
    Defined but not used: ‘f’
[ 4 of 35] Compiling Network.BitTorrent.Internal.Types ( src/Network/BitTorrent/Internal/Types.hs, dist/build/Network/BitTorrent/Internal/Types.o )
[ 5 of 35] Compiling Network.BitTorrent.Address ( src/Network/BitTorrent/Address.hs, dist/build/Network/BitTorrent/Address.o )

src/Network/BitTorrent/Address.hs:680:53: Warning:
    In the use of ‘bitSize’ (imported from Data.Bits):
    Deprecated: "Use 'bitSizeMaybe' or 'finiteBitSize' instead"
[ 6 of 35] Compiling Network.BitTorrent.DHT.Token ( src/Network/BitTorrent/DHT/Token.hs, dist/build/Network/BitTorrent/DHT/Token.o )
[ 7 of 35] Compiling Data.Torrent     ( src/Data/Torrent.hs, dist/build/Data/Torrent.o )
[ 8 of 35] Compiling Network.BitTorrent.DHT.Routing ( src/Network/BitTorrent/DHT/Routing.hs, dist/build/Network/BitTorrent/DHT/Routing.o )
[ 9 of 35] Compiling Network.BitTorrent.DHT.ContactInfo ( src/Network/BitTorrent/DHT/ContactInfo.hs, dist/build/Network/BitTorrent/DHT/ContactInfo.o )
[10 of 35] Compiling Network.BitTorrent.DHT.Message ( src/Network/BitTorrent/DHT/Message.hs, dist/build/Network/BitTorrent/DHT/Message.o )

src/Network/BitTorrent/DHT/Message.hs:201:10:
    Illegal instance declaration for
      ‘KRPC (Query FindNode) (Response (NodeFound ip))’
      The liberal coverage condition fails in class ‘KRPC’
        for functional dependency: ‘req -> resp’
      Reason: lhs type ‘Query FindNode’
        does not determine rhs type ‘Response (NodeFound ip)’
    In the instance declaration for
      ‘KRPC (Query FindNode) (Response (NodeFound ip))’

src/Network/BitTorrent/DHT/Message.hs:262:10:
    Illegal instance declaration for
      ‘KRPC (Query GetPeers) (Response (GotPeers ip))’
      The liberal coverage condition fails in class ‘KRPC’
        for functional dependency: ‘req -> resp’
      Reason: lhs type ‘Query GetPeers’
        does not determine rhs type ‘Response (GotPeers ip)’
    In the instance declaration for
      ‘KRPC (Query GetPeers) (Response (GotPeers ip))’
builder for `/nix/store/8c9aijxnc8in3h5rz39lvxacvrfhqw46-haskell-bittorrent-ghc7.8.3-0.0.0.3-shared.drv' failed with exit code 1
error: build of `/nix/store/8c9aijxnc8in3h5rz39lvxacvrfhqw46-haskell-bittorrent-ghc7.8.3-0.0.0.3-shared.drv' failed
```

**Comment by Fuuzetsu** [comment]
Does not occur with GHC 7.6.3

**Comment by pxqr** [comment]
Do not really understand why this is working with GHC 7.6.3.

**Comment by Fuuzetsu** [comment]
It works with 7.6.3 because FunctionalDependencies were broken there. See https://ghc.haskell.org/trac/ghc/ticket/8634 for discussion on allowing the old behaviour with an extension.

```
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE UnicodeSyntax #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE UndecidableInstances #-}
module T where

data I a = I a
data A = A
data B = B
data F a b = F a b

class C1 a where

class C a b | a → b where
  m ∷ F a b

instance C1 a => C A (I a) where
  m = F A (I undefined)

instance C1 a => C B (I a) where
  m = F B (I undefined)
```

Above is a smaller example of what is accepted by 7.6.3 and not 7.8.3. The problem is that the LHS does not determine the RHS of the FunDep (`A` can't determine `(I a)` unless `a` also appears on the LHS which it does not). This was fixed in in 7.8.3 so your invalid KRPC instance is now caught as it should be: `(Query GetPeers)` can't determine `(Response (GotPeers ip))` without `ip` occuring on LHS which it isn't.

It does not seem however that the KRPC FunDep is actually needed: removing it only affects one place in the bittorrent test-suite but otherwise all code compiles the same. If you wish to keep the FunDep then you'll have to fix the instance.

**Comment by pxqr** [comment]
Thanks for explanation, didn't knew about this bug.
